File: mpp-proiect-csharp-culbec/Network/NetworkUtils/AbstractServer.cs
using System.Net;
using System.Net.Sockets;
using log4net;

namespace Network.NetworkUtils;

public abstract class AbstractServer
{
    private static readonly ILog Log = LogManager.GetLogger(typeof(AbstractServer));
    private TcpListener? _server;
    private readonly string _host;
    private readonly int _port;

    protected AbstractServer(string host, int port)
    {
        _host = host;
        _port = port;
    }

    public void Start()
    {
        try
        {
            Log.Info("Initializing the IP address...");
            var ipAddress = IPAddress.Parse(_host);

            Log.InfoFormat("Initializing with {0}:{1}", _host, _port);
            var endPoint = new IPEndPoint(ipAddress, _port);

            Log.InfoFormat("Initializing the server with {0}", endPoint);
            _server = new TcpListener(endPoint);

            Log.Info("Starting the server...");
            _server.Start();

            Log.Info("Server started on port " + _port);
            while (true)
            {
                Log.Info("Waiting for clients...");
                var client = _server.AcceptTcpClient();

                Log.Info("Client connected. Processing the request...");
                ProcessRequest(client);
            }
        }
        catch (Exception e)
        {
            Log.ErrorFormat("An exception occurred: {0}", e.Message);
        }
        finally
        {
            Log.InfoFormat("Stopping the server...");
            Stop();
        }
    }

    private void Stop()
    {
        if (_server == null)
        {
            Log.Warn("The server was not initialized!");
            return;
        }

        try
        {
            _server.Stop();
            Log.Info("Server stopped.");
        }
        catch (Exception e)
        {
            Log.ErrorFormat("Couldn't stop the server: {0}", e.Message);
        }
    }

    protected abstract void ProcessRequest(TcpClient client);
}

public abstract class AbstractConcurrentServer : AbstractServer
{
    private static readonly ILog Log = LogManager.GetLogger(typeof(AbstractConcurrentServer));
    protected AbstractConcurrentServer(string host, int port) : base(host, port)
    {
    }

    protected override void ProcessRequest(TcpClient client)
    {
        Log.Info("Creating a worker...");
        var thread = CreateWorker(client);

        Log.Info("Starting the worker...");
        thread.Start();
    }

    protected abstract Thread CreateWorker(TcpClient client);
}

File: mpp-proiect-java-culbec/MotorcycleContest/Networking/src/main/java/NetworkUtils/AbstractServer.java
package NetworkUtils;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;

public abstract class AbstractServer {
    protected final Logger logger;
    private static final int DEFAULT_BACKLOG = 10;

    private final String hostname;
    private final int port;
    private ServerSocket server = null;

    public AbstractServer(String hostname, int port) {
        this.hostname = hostname;
        this.port = port;
        this.logger = LogManager.getLogger(this.getClass());
    }

    /**
     * Processes a request of a client.
     * @param client Client that sent the request.
     * @throws ServerException If the server encountered a problem on processing the request.
     */
    protected abstract void processRequest(Socket client) throws ServerException;

    /**
     * Starts the server.
     * @throws ServerException If the server encountered a problem on starting.
     */
    public void start() throws ServerException {
        logger.traceEntry("Starting the server...");
        try {
            // Creating a server socket.
            InetAddress address = Inet4Address.getByName(this.hostname);
            this.server = new ServerSocket(port, DEFAULT_BACKLOG, address);

            logger.info("Server started on port: {}", port);

            // Accepting clients indefinitely.
            while (true) {
                Socket client = server.accept();
                logger.info("Client connected: {}", client.getInetAddress());

                logger.info("Processing the request...");
                this.processRequest(client);
            }
        } catch (IOException e) {
            logger.error("Couldn't start the server: {}", e.getMessage());
            throw new ServerException(e.getMessage());
        } finally {
            // Stopping the server after crashing or after explicitly stopping it.
            this.stop();
        }
    }

    /**
     * Stops the server.
     * @throws ServerException If the server encountered a problem on stopping.
     */
    public void stop() throws ServerException {
        logger.traceEntry("Stopping the server...");
        try {
            // Closing the server socket.
            this.server.close();
        } catch (IOException e) {
            logger.error("Couldn't stop the server: {}", e.getMessage());
            throw new ServerException(e.getMessage());
        }
    }
}
